Refactor image caching

kurhadeprashant opened the pull request:
I refactored image caching. For big image I have scale down the image size and store it on disk.

cc @aleffert

aleffert commented on edXVideoLocker/ImageCache.m:
```diff
@@ -0,0 +1,182 @@
+@end
+
+
+@implementation ImageCache
```
This class should follow our naming guidelines - using the OEX name space.

kurhadeprashant replied:
Added OEX name convention for image class. 

cc @aleffert

aleffert commented on edXVideoLocker/ImageCache.m:
```diff
@@ -0,0 +1,182 @@
+
+@implementation ImageCache
+
++ (id)sharedInstance {
```
This says it is an general ImageCache, but actually it only works for a specific use case because the size of images it saves is hard coded.

kurhadeprashant replied:
Added NSCache support to get the image

aleffert commented on edXVideoLocker/ImageCache.m:
```diff
@@ -0,0 +1,182 @@
+    if (self = [super init]) {
+       // _imageCache =[[NSCache alloc]init];
+        self.imageQueue = [[NSOperationQueue alloc] init];
+        self.imageQueue.maxConcurrentOperationCount = 4;
```
Why ``4`?. If it's a magic number obtained from empirical performance testing it's good to leave that as a comment. I haven't tried this particular code, but I'll note that different devices will probably want a different number of concurrent operations and it's generally better to not set this at all and just let the system manage it.

kurhadeprashant replied:
Removed hard coded value for max concurrent operation count. 

cc @aleffert

aleffert commented on edXVideoLocker/ImageCache.m:
```diff
@@ -0,0 +1,182 @@
+
+@interface ImageCache()
+{
+ //   NSCache *_imageCache;
```
I'm not sure why this is commented out. Having an image cache backed by an NSCache makes a lot of sense. Given that we use this in a table view, going to disk every time seems like it will create performance problems.

kurhadeprashant replied:
We have to give support for offline mode as well that why we are keeping images on disk. Now we are first checking if images are there in NSCache if its not there then we are checking in to Disk if its there and if we don't find in Cache and Disk then we are fetching it form server.

aleffert commented on edXVideoLocker/ImageCache.h:
```diff
@@ -0,0 +1,19 @@
+{
+}
+@property (nonatomic, strong) NSOperationQueue *imageQueue;
++ (id)sharedInstance;
```
This should use ``instancetype`` not ``id``.

kurhadeprashant replied:
Changed id to instancetype 

cc @aleffert

aleffert commented on edXVideoLocker/ImageCache.h:
```diff
@@ -0,0 +1,19 @@
+
+#import <Foundation/Foundation.h>
+
+@interface ImageCache : NSObject
```
When creating a new singleton please have it be instantiated by OEXEnvironment (following the model of OEXConfig and OEXRouter). This will make it way easier to stub for tests.

aleffert commented on edXVideoLocker/ImageCache.h:
```diff
@@ -0,0 +1,19 @@
++ (id)sharedInstance;
+-(UIImage *)getImage:(NSString *)imageURLString;
+-(UIImage *)getImageFromCacheFromKey:(NSString *)imageURLKey;
+-(UIImage *)getDiskImage:(NSString *)filePath;
```
This is only used internally to the class so it shouldn't be part of the public interface.

kurhadeprashant replied:
Changed the implementation now we have only one method which is responsible for getting image with block. 

cc @aleffert

aleffert commented on edXVideoLocker/ImageCache.h:
```diff
@@ -0,0 +1,19 @@
+}
+@property (nonatomic, strong) NSOperationQueue *imageQueue;
++ (id)sharedInstance;
+-(UIImage *)getImage:(NSString *)imageURLString;
```
For something like this, it's better to expose an asynchronous interface. That way the burden of thread management is isolated into this one class instead of forced on every consumer of its interface.

For example
```
- (void)loadImageWithURL:(NSString*)imageURL completion:(void(^)(UIImage* image))completion;
```

kurhadeprashant replied:
Implemented one interface to get image. 

cc @aleffert

aleffert commented on edXVideoLocker/ImageCache.h:
```diff
@@ -0,0 +1,19 @@
+@property (nonatomic, strong) NSOperationQueue *imageQueue;
++ (id)sharedInstance;
+-(UIImage *)getImage:(NSString *)imageURLString;
+-(UIImage *)getImageFromCacheFromKey:(NSString *)imageURLKey;
```
This method appears to do nothing and should be removed.

kurhadeprashant replied:
Removed unnecessary methods. 

cc @aleffert

aleffert commented on edXVideoLocker/OEXMyVideosViewController.m:
```diff
@@ -569,36 +569,41 @@ - (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(N
+                    
+                    // get the UIImage
+                    
+                    UIImage *image = [imageCache getImage:imgURLString];
```
There is a race condition here. If you scroll quickly, the mapping between cells and courses may change and we'll end up with the wrong image on a course. We need to check that the course associated with the cell is the one associated with the image that loaded before setting the image.

This will be a lot simpler if the callback is inside the cell class. The block should save the current course_id and then the callback can compare the block with the cell's

kurhadeprashant replied:
Fixed this issue. 

cc @aleffert

aleffert commented on edXVideoLocker/OEXFrontCourseViewController.m:
```diff
@@ -23,10 +23,13 @@
 
 @interface OEXFrontCourseViewController ()
-
+{
```
Please use a property for this instead of an ivar. They have stronger documentation and make it easier to track where a property changes when debugging.

kurhadeprashant replied:
Removed ERROR_VIEW_HEIGHT as this is not used 

cc @aleffert

aleffert commented on edXVideoLocker/OEXImageCache.m:
```diff
@@ -0,0 +1,172 @@
+
+@interface OEXImageCache()
+{
+    NSCache *_imageCache;
```
Again, please use properties.

kurhadeprashant replied:
Added property for max file size 

cc @aleffert

aleffert commented on edXVideoLocker/OEXImageCache.m:
```diff
@@ -0,0 +1,172 @@
+    NSString * filePath = [OEXFileUtility completeFilePathForUrl:imageURLString];
+    UIImage *returnImage = nil;
+    if ([[NSFileManager defaultManager] fileExistsAtPath:filePath]) {
+        NSData * imageData = [[NSData alloc]initWithContentsOfFile:filePath];
```
The ``fileExistsAtPath:`` call here is unnecessary, easier to just try to load the image data and see if it returns ``nil``. Also you can use ``[UIImage imageWithContentsOfFile:]`` instead of loading the data and converting it to an image.

aleffert commented on edXVideoLocker/OEXImageCache.m:
```diff
@@ -0,0 +1,172 @@
+    UIGraphicsBeginImageContext(rect.size);
+    [image drawInRect:rect];
+    UIImage *img = UIGraphicsGetImageFromCurrentImageContext();
+    NSData *imageData = UIImageJPEGRepresentation(img, compressionQuality);
```
.5 is a pretty low quality jpeg. Have you done some comparisons to make sure we're not losing too much quality?

kurhadeprashant replied:
we have removed the quality constraint while compression we are keeping it as 1.0 which is don't compress image 

cc @aleffert

aleffert commented:
This needs unit tests. They can use a different path and delete it on initialization to ensure it doesn't interfere with any existing user data.

aleffert commented on edXVideoLocker/OEXImageCache.m:
```diff
@@ -0,0 +1,172 @@
+    float actualHeight = image.size.height;
+    float actualWidth = image.size.width;
+    float maxHeight = 500.0;
+    float maxWidth = 500.0;
```
I'm not that excited about hardcoding the sizes into this, but let's not worry too hard about it right now in the interesting of expedience. But we should at least do something appropriate to the current screen scale via ``[UIScreen mainScreen].scale``

kurhadeprashant replied:
Removed hard coded valued for maxHeight and maxWidth.

cc @aleffert

aleffert commented on edXVideoLocker/OEXImageCache.m:
```diff
@@ -0,0 +1,172 @@
+                    returnImage=[UIImage imageWithData:compressData];
+                    //write new file
+                    if (![compressData writeToFile:filePath atomically:YES]) {
+                        //ELog(@"There was a problem saving json to file");
```
Why is this log commented out? Also we should probably have an assert in this case?

aleffert commented on edXVideoLocker/OEXImageCache.m:
```diff
@@ -0,0 +1,172 @@
+-(UIImage *)getImage:(NSString *)imageURLString
+{
+    
+    NSString * filePath = [OEXFileUtility completeFilePathForUrl:imageURLString];
```
It seems weird to save the path for the URL with a compressed version. What if we need to use a different size for that same URL later. Let's append the size we're using to the path.

aleffert commented on edXVideoLocker/OEXImageCache.h:
```diff
@@ -0,0 +1,18 @@
+@property (nonatomic, strong) NSOperationQueue *imageQueue;
++ (instancetype)sharedInstance;
+-(UIImage *)getImage:(NSString *)imageURLString;
+-(UIImage *)getImageFromCacheFromKey:(NSString *)imageURLKey;
```
Is it a requirement to save the compressed version to disk? I'd be much happier if we didn't have to bother with that part. Then this would just be a nice little wrapper around NSCache that shrank images first.

aleffert commented on edXVideoLocker/OEXImageCache.h:
```diff
@@ -0,0 +1,17 @@
+@interface OEXImageCache : NSObject
+{
+}
+@property (nonatomic, strong) NSOperationQueue *imageQueue;
```
This property doesn't need to be public.

aleffert commented on edXVideoLocker/OEXImageCache.m:
```diff
@@ -0,0 +1,194 @@
+@interface OEXImageCache()
+{
+    NSCache *_imageCache;
+    CGFloat maxFileSize;
```
This is a constant, so it doesn't need an ivar.

Something like

```
static const CGFloat OEXImageCacheMaxFileBytes = 100 * 1024
```

aleffert commented:
We need to handle the case where a user logs out. Also, we should probably mark these as using NSURLIsExcludedFromBackupKey.

aleffert commented:
If I scroll up and down on a slow connection is this going to result in a lot of requests and compression operations for the same image?

kurhadeprashant commented:
Added NSURLIsExcludedFromBackupKey on image path. Also added filter for checking if image request is already in queue. it will avoid duplicate image requests

cc @aleffert

aleffert commented on edXVideoLocker.xcodeproj/project.pbxproj:
```diff
@@ -1906,6 +1914,7 @@
 			isa = XCBuildConfiguration;
 			baseConfigurationReference = 3BF001569FC2997AF924C889 /* Pods-edXVideoLocker.release.xcconfig */;
 			buildSettings = {
+				CODE_SIGN_IDENTITY = "iPhone Developer";
```
It looks like this is setting "iPhone Developer" on release builds, which probably isn't what we want.

aleffert commented on edXVideoLocker/OEXFrontCourseViewController.m:
```diff
@@ -393,44 +397,25 @@ - (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(N
+            [imageCache getImage:imgURLString completionBlock:^(UIImage *displayImage) {
+                if(displayImage){
+                    OEXFrontTableViewCell *cell = (OEXFrontTableViewCell *)[self.table_Courses cellForRowAtIndexPath:indexPath];
+                    if (cell && [cell isKindOfClass:[OEXFrontTableViewCell class]])
```
This still has a race condition if loading images is slow and the table gets refreshed.

We should save the current course or course id to a property on the cell and check if the cell's current course is the same as the one we set when we initiated the request.

kurhadeprashant replied:
Added check on course id before setting image for cell.

cc @aleffert

aleffert commented on edXVideoLocker/OEXFrontCourseViewController.m:
```diff
@@ -393,44 +397,25 @@ - (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(N
         
         OEXCourse *obj_course = [self.arr_CourseData objectAtIndex:indexPath.section];
-        
-        cell.course = obj_course;
```
Please don't delete this line. The ``newCourseContentClicked`` path uses the cell's course to determine which course to show.

kurhadeprashant replied:
Fixed 

cc @aleffert

aleffert commented on edXVideoLocker/OEXImageCache.m:
```diff
@@ -0,0 +1,229 @@
+-(void)removeAllObjectsFromCache
+{
+    [_imageCache removeAllObjects];
+    [_requestRecord removeAllObjects];
```
Do we need a way to remove the images on disk too or will that happen when we clear other saved URL data?

aleffert commented on edXVideoLocker/OEXImageCache.m:
```diff
@@ -0,0 +1,229 @@
+
+-(void)clearImageCache
+{
+    [self removeAllObjectsFromCache];
```
Since this only removes the parts of the cache that are in memory, the name should reflect that.

aleffert commented on edXVideoLocker/OEXImageCache.m:
```diff
@@ -0,0 +1,229 @@
+}
+
+- (void)dealloc {
+    // Should never be called, but just here for clarity really.
```
Not sure this comment is accurate. We might want to instantiate an instance of this with a shorter lifetime inside a unit test.

aleffert commented on edXVideoLocker/OEXImageCache.m:
```diff
@@ -0,0 +1,229 @@
+                {
+                    NSURL *imageURL=[NSURL URLWithString:imageURLString];
+                    {
+                        if([[_requestRecord valueForKey:imageURLString ]boolValue])
```
This should be ``objectForKey:`` not ``valueForKey:``.

aleffert commented on edXVideoLocker/OEXImageCache.m:
```diff
@@ -0,0 +1,229 @@
+                            completionBlock(nil);
+                            return;
+                        }
+                        [_requestRecord setValue:[NSNumber numberWithBool:YES] forKey:imageURLString];
```
You can just use ``@YES`` instead of ``[NSNumber numberWithBool:YES]``

aleffert replied:
None of the the accesses to _requestRecord here are thread safe. We could easily end up with multiple things writing to it at the same time, causing undefined behavior. We should probably update that on the main thread when first creating an operation. Then clear it, again on the main thread, at the end of the operation.

aleffert commented on edXVideoLocker/OEXImageCache.m:
```diff
@@ -0,0 +1,229 @@
+    }
+    return;
+}
+-(void)exludeiCloudImageBackup:(NSString *)path
```
``exludeiCloudImageBackup`` misspells ``exclude``

aleffert commented on edXVideoLocker/OEXMyVideosViewController.m:
```diff
@@ -569,37 +569,20 @@ - (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(N
-                    [_dataInterface downloadWithRequestString:[NSString stringWithFormat:@"%@%@", [OEXConfig sharedConfig].apiHostURL, obj_course.course_image_url]  forceUpdate:YES];
+            OEXImageCache *imageCache=[OEXImageCache sharedInstance];
+            [imageCache getImage:imgURLString completionBlock:^(UIImage *displayImage) {
+                if(displayImage){
```
This has the same race condition as above.

aleffert commented on edXVideoLocker/OEXImageCache.m:
```diff
@@ -0,0 +1,234 @@
+                    if([[_requestRecord objectForKey:imageURLString ]boolValue])
+                    {
+                        ELog(@"Duplicate image download request. Already in progress");
+                        completionBlock(nil);
```
Maybe I'm missing something, but won't this just cause the image request to fail? So if you queue up one and then another one, the second one will just be return nil?

kurhadeprashant replied:
Added NSError in block if request is duplicate. 

cc @aleffert

aleffert commented on edXVideoLocker/OEXMyVideosViewController.m:
```diff
@@ -563,43 +563,26 @@ - (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(N
         NSDictionary *dictVideo = [self.arr_CourseData objectAtIndex:indexPath.section];
         
-        OEXCourse *obj_course = [dictVideo objectForKey:CAV_KEY_COURSE];
+       __block OEXCourse *obj_course = [dictVideo objectForKey:CAV_KEY_COURSE];
```
We would only need ``__block`` if we changed this inside the block.

kurhadeprashant replied:
Fixed 

cc @aleffert

aleffert commented on edXVideoLockerTests/OEXImageCacheTests.m:
```diff
@@ -0,0 +1,102 @@
+    __block BOOL isComplete=NO;
+   
+    OEXImageCache *imageCache =[OEXImageCache sharedInstance];
+    [imageCache getImage:@"https://courses.edx.org/c4x/KIx/KIPractihx/asset/kix_pragmatic_course_banner608x211.jpg" completionBlock:^(UIImage *displayImage) {
```
Can we mock this or use a file on disk or something? The unit tests shouldn't depend on having a network connection.

kurhadeprashant replied:
Currently in this project we don't have OCMock integrated. Its there in main branch after merging we can change the test case. Let me know your view on this

cc @aleffert

aleffert replied:
Just use a ``file`` url referencing an image on disk, you can use ``NSBundle`` to get a path to an image in the bundle and then turn it into a ``file`` URL.

aleffert commented:
Sorry for the long back and forth. Just a couple more things.

aleffert commented on edXVideoLockerTests/OEXImageCacheTests.m:
```diff
@@ -0,0 +1,102 @@
+    }];
+    while (!isComplete) {
+        NSTimeInterval const interval = 0.002;
+        if (! [[NSRunLoop currentRunLoop] runMode:NSDefaultRunLoopMode beforeDate:[NSDate dateWithTimeIntervalSinceNow:interval]]) {
```
Can we use XCTestExpectation here?

kurhadeprashant replied:
Fixed 

cc @aleffert

aleffert commented on edXVideoLocker/OEXImageCache.m:
```diff
@@ -0,0 +1,240 @@
+                        [errorDetail setValue:@"Request is already in queue" forKey:NSLocalizedDescriptionKey];
+                        NSError *error = [NSError errorWithDomain:@"Forbidden" code:403 userInfo:errorDetail];
+                        
+                        completionBlock(nil,error);
```
Sorry. I guess I wasn't clear. We don't want to just send an error this case. It doesn't solve the real problem which is that scrolling down and up will result in no image, even though we're in the middle of loading the image. An easy solution to this problem is to have an ``NSNotification`` fire when the image loads. The ``object`` for the notification can be the URL and then the cell can just check if that URL matches its course image URL and if so, update its image.

kurhadeprashant replied:
Used NSNotification for image downloading. 

cc @aleffert

aleffert commented on edXVideoLockerTests/OEXImageCacheTests.m:
```diff
@@ -0,0 +1,105 @@
+@implementation OEXImageCacheTests
+
+- (void)setUp {
+    [super setUp];
```
unused code.

aleffert commented:
:+1:
